packages/core/snapshot/snapshotwrite.go

```diff
@@ -0,0 +1,184 @@
+	header *ledger.SnapshotHeader,
+	outputProd OutputWithMetadataProducerFunc,
+	epochDiffsProd EpochDiffProducerFunc) error {
+
+	writeFunc := func(name string, value any, offsetsToIncrease ...*int64) error {
+		return writeFunc(writeSeeker, name, value, offsetsToIncrease...)
+	}
+
+	writeOutputWithMetadatasFunc := func(chunks []*ledger.OutputWithMetadata) error {
+		if len(chunks) == 0 {
+			return nil
+		}
+
+		typeSet := new(serix.TypeSettings)
+		data, err := serix.DefaultAPI.Encode(context.Background(), chunks, serix.WithTypeSettings(typeSet.WithLengthPrefixType(serix.LengthPrefixTypeAsUint32)), serix.WithValidation())
+		if err != nil {
+			return err
+		}
+		if err := writeFunc("outputs", append(data, delimiter...)); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	err := writeSnapshotHeader(writeSeeker, header)
+	if err != nil {
+		return err
+	}
+
+	// write outputWithMetadata
+	var outputWithMetadataCounter uint64 = 0
+	var outputChunkCounter int
+	chunksOutputWithMetadata := make([]*ledger.OutputWithMetadata, 0)
+	for {
+		output := outputProd()
+		if output == nil {
+			// write rests of outputWithMetadatas
+			err = writeOutputWithMetadatasFunc(chunksOutputWithMetadata)
+			if err != nil {
+				return err
+			}
+			break
+		}
+
+		outputWithMetadataCounter++
+		chunksOutputWithMetadata = append(chunksOutputWithMetadata, output)
+
+		// put a delimeter every 100 outputs
+		if outputChunkCounter == 100 {
+			err = writeOutputWithMetadatasFunc(chunksOutputWithMetadata)
+			if err != nil {
+				return err
+			}
+			chunksOutputWithMetadata = make([]*ledger.OutputWithMetadata, 0)
+		}
+	}
+
+	// write epochDiffs
+	epochDiffs, err := epochDiffsProd()
+	if err != nil {
+		return err
+	}
+
+	typeSet := new(serix.TypeSettings)
+	bytes, err := serix.DefaultAPI.Encode(context.Background(), epochDiffs, serix.WithTypeSettings(typeSet.WithLengthPrefixType(serix.LengthPrefixTypeAsUint32)), serix.WithValidation())
+	if err != nil {
+		return err
+	}
+	if err := writeFunc(fmt.Sprintf("diffEpoch"), append(bytes, delimiter...)); err != nil {
+		return err
+	}
+
+	// seek back to the file position of the outputWithMetadata counter
+	if _, err := writeSeeker.Seek(0, io.SeekStart); err != nil {
+		return fmt.Errorf("unable to seek to LS counter placeholders: %w", err)
+	}
+	if err := writeFunc(fmt.Sprintf("outputWithMetadata counter %d", outputWithMetadataCounter), outputWithMetadataCounter); err != nil {
+		return err
+	}
+	header.OutputWithMetadataCount = outputWithMetadataCounter
+
+	return nil
+}
+
+// NewLedgerOutputWithMetadataProducer returns a OutputWithMetadataProducerFunc that provide OutputWithMetadatas from the ledger.
+func NewLedgerOutputWithMetadataProducer(l *ledger.Ledger) OutputWithMetadataProducerFunc {
+	prodChan := make(chan interface{})
+
+	go func() {
+		l.ForEachAcceptedUnSpentOutputWithMetadata(func(o *ledger.OutputWithMetadata) {
+			prodChan <- o
+		})
+
+		close(prodChan)
+	}()
+
+	binder := producerFromChannels(prodChan)
```

Comment:
Do we need an extra function if it is only used here? I feel it makes the code harder to follow